ocrd/ocrd/cli/workspace.py

```diff
@@ -159,6 +167,54 @@ def workspace_add_file(ctx, file_grp, file_id, mimetype, page_id, force, local_f
     )
     workspace.save_mets()
 
+# ----------------------------------------------------------------------
+# ocrd workspace add-group
+# ----------------------------------------------------------------------
+
+@workspace_cli.command('add-group')
+@click.option('-m', '--mimetype', help="Media type of the file", required=True)
+@click.option('-f', '--force', help="Continue even if mets:file same ID already exists", default=False, is_flag=True)
+@click.argument('file_grp_dir', type=click.Path(dir_okay=True, file_okay=False, readable=True, resolve_path=True), required=True)
+@pass_workspace
+def workspace_add_file_group(ctx, mimetype, force, file_grp_dir):
+    """
+    Add a directory of files as a fileGrp to METS.
+
+    Convention:
+        FILEGRP_DIR must reside within the workspacee
+        fileGrp ID = last folder name component
+        page ID = filename sans extension
+        file ID = sanitized concatenation of fileGrp ID and page ID
+    """
+    workspace = Workspace(ctx.resolver, directory=ctx.directory, mets_basename=ctx.mets_basename, automatic_backup=ctx.automatic_backup)
+
+    file_grp_dir = Path(file_grp_dir).resolve()
+
+    ws_dir = Path(ctx.directory).resolve()
+    print(file_grp_dir)
+
+    # XXX will raise ValueError if file_grp_dir isn't relative to ctx.directory
+    file_grp_dir.relative_to(ws_dir)
+    file_grp = clean_id(file_grp_dir.name)
+
+    for f in file_grp_dir.iterdir():
+        page_id = f.name
+        if f.suffix:
+            page_id = page_id[0:-len(f.suffix)]
```

Comment:
Shouldn't we at least try to map these onto existing pages in the structMap? You could try to (remove a common prefix and) find substrings shared by existing pageIds systematically.

Scenario: `add-group` for images followed by `add-group` for PAGE-XML files.

Reply:
> Scenario: add-group for images followed by add-group for PAGE-XML files.

The assumption here is that file (base) names would be the same for different file groups. But I agree that this assumption is too restrictive. I think the convention of many libraries seems to be to have a file-type specific suffix to the file ID (e.g. `FILE_0009_THUMBS`) and page IDs that begin with `PHYS`, such as `PHYS_0009`. So it gets tricky fast with a common convention.

How about allowing users to define regular expressions for searching and template strings for creating ID on the command line. Something like

```
ocrd workspace add-group --file-id="s/.*(\d{4}).*/FILE_\1_DEFAULT" --page-id="s/.*(\d{4}).*/PHYS_\1/" ...
```
```

Reply:
Yes, that would be a good start. Regexps are complicated but easy and safe to parse.

But I think we should _also_ offer some kind of default which guesses these as best as we can.

Reply:
I've rewritten the implementation. Here's an example how this can be used:

Let's assume you have directories `path/to/files/PAGE/` containing PAGE files and `path/to/files/IMG` with images. The files have a basename `page_0001.xml`, ``page_0001.tif'` etc.

```sh
ocrd workspace bulk-add \
    --regex '^.*/(?P<fileGrp>[^/]+)/page_(?P<pageid>.*)\.(?P<ext>[^\.]*)$' \
    --file-id 'FILE_{{ fileGrp }}_{{ pageid }}' \
    --page-id 'PHYS_{{ pageid }}' \
    --file-grp "{{ fileGrp }}" \
    --url '{{ fileGrp }}/FILE_{{ pageid }}.{{ ext }}' \
    'path/to/files/*/*.*'
```

This will first expand the glob to get filenames and resolve them to absolute paths.

Every path is then matched against `--regex` with `re.match`, yielding template variables derived from the syntax of the path. These template variables can be used in all file-specific options. `--url` after expansion is used as the filename relative to the workspace directory and copied into the workspace if not already present. After expanding all template variables, the file is added with `Workspace.add_file`.

In this case:
  * `path/to/files/PAGE/page_0001.xml` ->
    * `url`: `PAGE/FILE_0001.xml` (which will be copied unless `path/to/files` **is** the workspace dir)
    * `fileGrp`: `PAGE`
    * `ID`: `FILE_0001`
    * `pageId`: `PHYS_0001`

`--mimetype`, if not provided, is mapped from the file extension.

`--ignore` will disable the check for existing files with the same `@ID` and is a huge performance boost.

It may not be the prettiest solution but it is very versatile, as long as all the information can be derived directly from the file path. If the process is more involved, e.g. using mappings or string transformation methods, we could add a proper template engine like jinja. On the other hand, bulk renaming files with unix tools and fine-tuning the regex is probably easier.

We could also offer or at least document ready-made regex/template combinations for common use cases.

Reply:
This is an excellent idea and solution. It has maximum flexibility but has a clear interface and simple implementation, and is thus most reliable.

>     * (which will be copied unless `path/to/files` **is** the workspace dir)

Not quite: IIUC, files will be copied unless they exist exactly _under (expanded) `url`_ in the workspace directory. So in your example, because `regex` differs from `url`, they will be copied even if `path/to/files` is the workdir.

> On the other hand, bulk renaming files with unix tools and fine-tuning the regex is probably easier.

I agree: more elaborate schemes can easily be constructed with a simple (and fast) shell loop. `bulk-add` only needs to be the last step. 

But there's a catch: I believe one still needs an additional option to allow **skipping** files that don't match the regex. This could be for another run of `bulk-add`, or just for files that we don't care about (presentation, logs, meta-data etc).

For example, to import the old (first-generation zip-file) OCR-D GT directories, one could then do:
```
ocrd workspace bulk-add --skip --regex '^.*/(?P<dispname>[^/]*)/(?P=dispname)_(?P<pageid>[0-9]*)\.tif$' --file-id 'FILE_ORIG_{{ pageid }}' --page-id 'PHYS_{{ pageid }}' --file-grp OCR-D-IMG --url '{{ dispname }}_{{ pageid }}.tif' # only images, no copying
ocrd workspace bulk-add --skip --regex '^.*/(?P<dispname>[^/]*)/page/(?P=dispname)_(?P<pageid>[0-9]*)\.xml$' --file-id 'FILE_GT_{{ pageid }}' --page-id 'PHYS_{{ pageid }}' --file-grp OCR-D-GT-SEG-PAGE --url 'page/{{ dispname }}_{{ pageid }}.xml' # only PAGE, no copying
ocrd workspace bulk-add --skip --regex '^.*/(?P<dispname>[^/]*)/alto/(?P=dispname)_(?P<pageid>[0-9]*)\.xml$' --file-id 'FILE_GT-ALTO_{{ pageid }}' --page-id 'PHYS_{{ pageid }}' --file-grp OCR-D-GT-ALTO-SEG-PAGE --mimetype application/alto+xml --url 'alto/{{ dispname }}_{{ pageid }}.xml' # only ALTO, no copying
```
(You cannot match the non-existing image subdirectory as fileGrp in this convention directly, and breaking it up allows a basic form of string transformation.)

Reply:
`--skip` does that now, thx to your suggestion in dac4b9f5

Reply:
Where would be a good place to document recipes like that – maybe in the setup guide or workflow guide?

Reply:
Both make sense, though the information would get a bit lost in the workflow guide. For now, I've [created a page on bulk-add in the ocrd-website wiki](https://github.com/OCR-D/ocrd-website/wiki/Workspace-Bulk-Add) with the recipes we've discussed here.